File: src/Test/ChurchFFI.fs
module Test.ChurchFFI

type Church<'a> = ('a -> 'a) -> 'a -> 'a
let zero : Church<'a> = fun _ value -> value
let successor (previous: Church<'a>) : Church<'a> = fun f value -> f (previous f value)
let rec fromInt n : Church<int> =
    if n = 0 then zero else successor (fromInt (n - 1))
let multiply (m: Church<'a>) (n: Church<'a>) : Church<'a> = fun f value -> m (n f) value
let square n = multiply (fromInt n) (fromInt n)
let runChurchFFI (input: obj) =
    let n = unbox<int> input
    let fourthPower = multiply (square n) (square n)
    let fifthPower = multiply fourthPower (fromInt n)
    fifthPower ((+) 1) 0 :> obj

File: src/Test/ChurchFFI.java
@FunctionalInterface
private interface Church<A> {
    java.util.function.Function<A, A> apply(java.util.function.Function<A, A> f);
}

private static <A> Church<A> zero() { return f -> x -> x; }
private static <A> Church<A> successor(Church<A> previous) {
    return f -> x -> f.apply(previous.apply(f).apply(x));
}

private static Church<Integer> fromInt(int n) {
    return n == 0 ? zero() : successor(fromInt(n - 1));
}

private static <A> Church<A> multiply(Church<A> m, Church<A> n) {
    return f -> x -> m.apply(n.apply(f)).apply(x);
}

private static Church<Integer> square(int n) {
    return multiply(fromInt(n), fromInt(n));
}

public static final java.util.function.Function<Object, Object> runChurchFFI = Bench.nativeBenchmark(input -> {
    int n = ((Number) input).intValue();
    Church<Integer> fourthPower = multiply(square(n), square(n));
    Church<Integer> fifthPower = multiply(fourthPower, fromInt(n));
    return fifthPower.apply(x -> x + 1).apply(0);
});
